bin/fuzz/fuzz_targets/fork-tree.rs

```diff
@@ -0,0 +1,166 @@
+        // In order to generate a valid list of operations, we must track the tree separately.
+        // To do so, we use `Rc`s and `Weak`s, which is the easiest (but non optimal) way to
+        // maintain a tree.
+
+        struct Node {
+            id: Cell<Option<u32>>, // `None` iff root.
+            parent: Weak<Node>,
+            children: RefCell<Vec<Rc<Node>>>,
+        }
+
+        impl Node {
+            fn gather_children(&self, out: &mut Vec<Rc<Node>>) {
+                let children = self.children.borrow();
+                out.extend(children.iter().cloned());
+                for child in children.iter() {
+                    child.gather_children(out);
+                }
+            }
+
+            fn prune_uncles(self: &Rc<Self>) {
+                if let Some(parent) = self.parent.upgrade() {
+                    let mut parent_children = parent.children.borrow_mut();
+                    parent_children.retain(|c| Rc::ptr_eq(c, self));
+                    parent.prune_uncles();
+                }
+            }
+        }
+
+        let mut operations = Vec::new();
+
+        let mut tree_root = Rc::new(Node {
+            id: Cell::new(None),
+            parent: Weak::new(),
+            children: RefCell::new(Vec::new()),
+        });
+
+        let mut next_node_id = 0;
+
+        u.arbitrary_loop(None, None, |u| {
+            let mut all_nodes = Vec::new();
+            all_nodes.push(tree_root.clone());
+            tree_root.gather_children(&mut all_nodes);
+
+            // Very rarely, we simply clear the tree.
+            if u.ratio(1, 30)? {
+                operations.push(Operation::Clear);
+                tree_root = Rc::new(Node {
+                    id: Cell::new(None),
+                    parent: Weak::new(),
+                    children: RefCell::new(Vec::new()),
+                });
+
+                return Ok(ControlFlow::Continue(()));
+            }
+
+            // 9/10th of the time, insert a new node in the tree.
+            if u.ratio(9, 10)? {
+                let index = u.choose_index(all_nodes.len())?;
+
+                operations.push(Operation::Insert {
+                    parent: all_nodes[index].id.get(),
+                    new_id: next_node_id,
+                });
+                let parent = Rc::downgrade(&all_nodes[index]);
+                all_nodes[index].children.borrow_mut().push(Rc::new(Node {
+                    id: Cell::new(Some(next_node_id)),
+                    parent,
+                    children: RefCell::new(Vec::new()),
+                }));
+                next_node_id += 1;
+
+                return Ok(ControlFlow::Continue(()));
+            }
+
+            // The remaining half of the time, we prune ancestors. Otherwise, we prune uncles.
+            if u.ratio(1, 2)? {
+                let index = u.choose_index(all_nodes.len() - 1)? + 1; // Avoid the tree root
```

Comment:
this will fail if the tree only consists of the single root element, right? It's fine, just something to be aware of if it's the case indeed.

Reply:
Yeah, `choose_index(0)` returns an error, which then instructs the fuzzer to change the data.